### PR TITLE
Rein in the udev rules file to fix bug #1526418

### PR DESCRIPTION
**Pull request by Pegasus-RPG**
This version restricts USB access to HID class devices and explicitly allows access to all USB devices from vendors that make known USB-bulk controllers.

### CODE REVIEW

**Comment by ywwg** [comment]
I'm going to test this with the HID devices I have, but I can't do that until later tonight.  Do we want to gate RC2 on this fix?  

(and  too bad I had to return the S4 or I'd check that as well. )

**Comment by Pegasus-RPG** [comment]
Yes, since it'll only take a minute to test this.

**Comment by ywwg** [comment]
so what I'd like to say is:  if testing goes well, commit and take the first build that includes this and the windows test fix.  If this PR has issues, just release RC2 anyway and we'll work on this more for the release.

**Comment by Pegasus-RPG** [comment]
Sounds good

**Comment by Be-ing** [comment]
IMO https://bugs.launchpad.net/mixxx/+bug/1517738 should be fixed before releasing RC2 so the udev rule can actually be tested.

**Comment by Pegasus-RPG** [comment]
Good point, but if that's an issue, how did the reporter of https://bugs.launchpad.net/mixxx/+bug/1526418 even have the problem?

**Comment by Be-ing** [comment]
Because they run Arch, not Ubuntu.

**Comment by Be-ing** [comment]
I don't think the rule should give permission for all HID devices. If someone really wants to make an HID mapping for their touchpad, they should be smart enough to add an line to this file or at least ask for help to do so. For most users though, seeing their mouse and keyboard listed along side their DJ controller is weird and likely cause frustration when they unknowingly disable normal functionality of their mouse or keyboard by enabling it as an HID controller in Mixxx. FWIW, I don't think anyone has made an HID mapping for a touchpad (if they have, they haven't mentioned it on the forums).

**Comment by Pegasus-RPG** [comment]
See, this is a tough one. There is a use case to have a second touch pad and/or keyboard (even one of the small numeric keypads) specifically to map to Mixxx functions (such as the touch pad on the VMS4.) I don't want to artificially prohibit using those kinds of things. How many people have gotten confused with this on the forums or in bug reports?

**Comment by Pegasus-RPG** [comment]
Also, limiting to specific vendors will result in questions and bug reports when a controller from a new vendor appears and it doesn't show up in Mixxx.

**Comment by Be-ing** [comment]
I haven't seen a single user ask about mapping a touchpad or keyboard with HID. A better solution than creating a situation that is confusing and fairly likely to cause issues for most users is to write HID mapping documentation and explain how to modify the udev rule for new devices.

File: res/linux/15-mixxx-usb.rules [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# This udev rule allows Mixxx to access HID and USB Bulk controllers when running as a normal user
```
Comment (rryan) [inline res/linux/15-mixxx-usb.rules]: Reading up on Debian packaging for udev rules, we aren't supposed to put the priority in the name. Could you rename this file ```mixxx.usb.rules```?

Comment (Pegasus-RPG) [inline res/linux/15-mixxx-usb.rules]: I had a feeling...

Comment (rryan) [inline res/linux/15-mixxx-usb.rules]: To work with dh_installudev which I believe is what we're supposed to be using, it needs to be named ```mixxx.usb.rules``` not ```mixxx-usb.rules```.

http://manpages.ubuntu.com/manpages/intrepid/man1/dh_installudev.1.html

**Comment by Be-ing** [comment]
@rryan what do you think about enabling all HID devices vs targeting specific vendors?

**Comment by Be-ing** [comment]
Here is what I propose for the rules. Can people with HID devices test this?
```
# This udev rule allows Mixxx to access HID controllers when running as a normal user

# Allow communicating with all USB devices made by Hercules. This includes HID as well as USB bulk controllers.
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ATTRS{idVendor}=="06f8", GROUP="users", MODE="0660"

# Allow communicating with HID devices
ATTRS{idVendor}=="08e4", ATTRS{bInterfaceClass}=="03", GROUP="users", MODE="0660" # Pioneer
ATTRS{idVendor}=="17cc", ATTRS{bInterfaceClass}=="03", GROUP="users", MODE="0660" # Native Instruments
ATTRS{idVendor}=="1157", ATTRS{bInterfaceClass}=="03", GROUP="users", MODE="0660" # Eks
ATTRS{idVendor}=="057e", ATTRS{bInterfaceClass}=="03", GROUP="users", MODE="0660" # Nintendo
ATTRS{idVendor}=="054c", ATTRS{bInterfaceClass}=="03", GROUP="users", MODE="0660" # Sony
```

**Comment by Pegasus-RPG** [comment]
In general Be, we prefer to allow maximum functionality and future expandability wherever possible. (It's one of the key things that sets Mixxx apart.) The only exception would be if doing so actually caused user confusion (like with the MIDI Through device on ALSA.) So I would say to leave it open now, and we can always limit it later if we do actually get complaints. (As I said above, this also lessens our maintenance burden.)

**Comment by Be-ing** [comment]
Thinking back, targeting specific vendor IDs with the first version of the udev rule was motivated by security concern. What are the security implications of allowing users to access all USB HID devices?

**Comment by ywwg** [comment]
I have definitely heard of DJs who map the touchpad to fx knobs.  And I don't think we should give our users needless busy work like editing an obscure file if we don't have to.  But we do have to be sure we're not undermining a distro's attempt to preserve security.

Maybe instead of "users" we can find a group with an equivalent permission level. For instance on Ubuntu and Debian, the "plugdev" group is allowed to mount external storage devices.  Or, the "audio" group has the right to use external and internal audio devices.  Maybe one of those is a better fit than the generic "users" group.

**Comment by Be-ing** [comment]
Maybe we can leave the group as "users" in the Mixxx source tree and leave it to packagers to replace it with a more appropriate group because every distro does it differently.

**Comment by Pegasus-RPG** [comment]
I've asked the Debian Mixxx package maintainer for his thoughts.

**Comment by daschuer** [comment]
It would be nice if one find the time to test  
https://github.com/mixxxdj/mixxx/pull/812
and/or 
https://github.com/mixxxdj/mixxx/pull/790
before RC2 as well. 
No release blockers or any reason for an extra delay of cause, but I suffer these issues by my selves. 
Thank you.

**Comment by MK-42** [comment]
I've tested both versions, the current one of this PR and  @Be-ings  from the comment above.
Both work fine with my usb printer.

**Comment by Be-ing** [comment]
@MK-42 thanks for testing.

**Comment by rryan** [comment]
Had you decided on priority 15 for some reason? It looks like the default is 40.

**Comment by rryan** [comment]
IMO we should warn the user in Mixxx if they're turning on a mouse or keyboard that enabling that device could steal the device from the OS. Or decide inside of Mixxx whether to expose a keyboard/mouse device rather than making that choice at the udev level. 

I think it's probably a better user experience to show more devices rather than fewer. HID devices not showing in Mixxx is a common stumbling point (since we have a FAQ for it).

Security-wise, I'm not entirely clear what threat models we're discussing here so it would be nice to make them explicit. Are we thinking of the "computer lab" situation where students would gain access to other, non-musical devices on the system that they otherwise would not have been able to access?

**Comment by rryan** [comment]
Worth noting -- our Debian maintainer doesn't even use our packaging. The debian packaging material in our tree is purely for our PPA and build-server produced debs.

**Comment by ywwg** [comment]
as a quick fix can we blacklist mice and keyboards?

**Comment by rryan** [comment]
@Pegasus-RPG -- my latest fix in 1.12 for packaging the udev rules uses the name as "res/linux/mixxx-usb-hid.rules". It needs to be a ```.udev``` file in the debian folder so I actually change the name when I copy it to the Debian folder. 

So you can name this whatever you want, just make sure to update the copy command:
https://github.com/mixxxdj/mixxx/blob/1.12/src/SConscript#L757

**Comment by Pegasus-RPG** [comment]
I got priority 15 from some article on the topic when I originally added the material to the Wiki. I don't remember now why that particular number. The bug poster did say that it comes before the hot-plugged USB printers though so therefore doesn't cause a problem. In order to make sure the bug is fixed though, we should add a low priority number to the file when its 'scons install'ed.

**Comment by Be-ing** [comment]
> as a quick fix can we blacklist mice and keyboards?

As a quick fix, no, because we'd need to come up with a method of detecting whether an HID device is a mouse or keyboard and test it thoroughly. I think a white list approach would be more reliable.

**Comment by ywwg** [comment]
> IMO we should warn the user in Mixxx if they're turning on a mouse or keyboard that enabling that device could steal the device from the OS

I don't think popping up a warning is feasible for this RC for sure.  That seems to be the only concrete example of a bad outcome from blindly displaying all HID devices.  As for threat models etc, I don't have anything in mind.  The original bug was more of an issue with break an OS than a specific security issue.  But if an OS has chosen not to allow users default access to HID, I'm not sure it's mixxx's place to override that decision.

**Comment by rryan** [comment]
Hm, we do have a keyboard and trackpad HID preset in the tree for the Apple bluetooth keyboard/touchpads.

**Comment by rryan** [comment]
> But if an OS has chosen not to allow users default access to HID, I'm not sure it's mixxx's place to override that decision.

I think this is purely about our PPA and self-made Debian packages. Distros -- even Debian, do their own thing here and if they're unhappy with what we've done I think they'll edit it. Our Debian package doesn't use this code so we're going to have to talk with Alessio about updating the Debian package to include these rules if we want the Debian-provided package to come with these.

I don't really have strong opinions on what to do here though.

**Comment by ywwg** [comment]
I think for now we should stick with what we have -- a relatively blanket policy that does expose mice to the list.  I've never seen a report of a user disabling their mouse by enabling it as a controller so I don't see a reason to block the release based on that hypothetical.  I also don't want to take the risk of making this file more restrictive and possibly blocking existing HID controllers from working.  This relatively minor change fixes the bug in question and I don't see a compelling reason to do any more at this point.

**Comment by Be-ing** [comment]
I don't think the issue of accidentally disabiling a mouse or keyboard has come up because the udev rule file has not previously been shipped with Mixxx. Until now, only users who have HID controllers have put the udev rule on the wiki on their system--which is a minority of users. The risk of a user accidentally disabiling their mouse or keyboard will be much greater now that all GNU/Linux users will have the udev rule on their system.

**Comment by rryan** [comment]
> On the other hand, I did check and saw that mice appear on my roommates' Windows computer (I don't know about Mac) and I haven't seen any Windows users complain.

On Mac mice and keyboards show up too.

**Comment by ywwg** [comment]
We think this is a decent incremental change, so we're going to start with this for now.

**Comment by Be-ing** [comment]
Yeah, considering that the issue hasn't come up with any Windows or Mac users, I think this is okay for now. However, I think it'd be good for 2.1 to come up with a way to warn users about enabling mice and keyboards as HID controllers. I propose checking the USB vendor ID against a white list of HID DJ controller manufacturers and warning if the HID device does not match that.

**Comment by ywwg** [comment]
we will definitely keep working on this issue, including considering all of the suggestions in this pr

**Pull request merged**